### PR TITLE
Configure contextual and structured logging for NEG controller

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
This PR incorporates changes suggested by structured logging ([KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/1602-structured-logging/README.md)) and contextual logging ([KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/3077-contextual-logging/README.md)) into the NEG controller, which improves the overall observability of the logs.

Migration guide that was referred: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

Changes introduced:
* Update klog version from v1 to v2
* All NEG controller log messages will be prefixed by `NEGController`
  Example:
  ```
  "NEGController: Starting network endpoint group controller"
  ```
* Various sub-components like `Syncer`, `ReadinessReflector`, `ClusterL4ILBEndpointsCalculator` etc will also have their component name prefixed.
  Example:
  ```
  "NEGController/Syncer: Stopping NEG syncer for service port"
  ...
  "NEGController/ReadinessReflector: Starting NEG readiness reflector"
  ```
* Syncer (as well as other some other sub-component logs), will include additional context like the name of the service and NEG.
  Example:
  ```
  "NEGController/Syncer: Stats for NEG" service="default/neg-demo-svc" negName="my-neg-name" description="desired NEG endpoints" stats="3 endpoints in zone \"us-central1-c\""
  ```

### CODE REVIEW

**Comment by gauravkghildiyal** [comment]
/assign @swetharepakula 

It would make sense to review the commits separately. The delta looks large, but the major contributor in this delta is the changes to the vendor directory introduced in the first commit _"Update klog version to v2"_.  The second commit _"Add contextual logging to NEG controller and its components"_ is the one which includes the changes to the logs.

File: pkg/neg/syncers/transaction.go [hunk 1]
```diff
@@ -545,12 +551,12 @@ func (s *transactionSyncer) logStats(endpointMap map[string]negtypes.NetworkEndp
 	for zone, endpointSet := range endpointMap {
 		stats = append(stats, fmt.Sprintf("%d endpoints in zone %q", endpointSet.Len(), zone))
 	}
-	klog.V(3).Infof("For NEG %q, %s: %s.", s.NegSyncerKey.NegName, desc, strings.Join(stats, ","))
+	s.logger.V(3).Info("Stats for NEG", "description", desc, "stats", strings.Join(stats, ","))
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:554]: Perhaps we can do something like this:
```
        stats := []string{}
	for zone, endpointSet := range endpointMap {
		stats = append(stats, zone, fmt.Sprintf("%d endpoints", endpointSet.Len()))
	}
	s.logger.V(3).Info("Endpoint stats for NEG", "description", desc, stats...)
```

**Review by swetharepakula — COMMENTED** [review]
Thanks Gaurav! Comments are mostly on improving the log level of log lines and some suggestions to make the log line clearer.

File: pkg/neg/syncers/transaction.go (hunk 1 above)
Comment (gauravkghildiyal) [reply pkg/neg/syncers/transaction.go]: Thanks. I've tweaked things to get it to compile. Let me know if it looks fine.

File: pkg/neg/manager.go [hunk 2]
```diff
@@ -539,7 +544,7 @@ func (manager *syncerManager) garbageCollectNEGWithCRD() error {
 	for _, cr := range deletionCandidates {
 		shouldDeleteNegCR := true
 		deleteByZone := len(cr.Status.NetworkEndpointGroups) == 0
-		klog.V(2).Infof("Deletion candidate %s/%s has %d NEG references", cr.Namespace, cr.Name, len(cr.Status.NetworkEndpointGroups))
+		manager.logger.V(2).Info(fmt.Sprintf("Deletion candidate has %d NEG references", len(cr.Status.NetworkEndpointGroups)), "svcneg", klog.KObj(cr))
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/manager.go]: Done!

File: pkg/neg/syncers/transaction.go [hunk 3]
```diff
@@ -256,17 +262,17 @@ func (s *transactionSyncer) syncInternalImpl() error {
...
-	filterEndpointByTransaction(committedEndpoints, s.transactions)
+	filterEndpointByTransaction(committedEndpoints, s.transactions, s.logger)
 
 	if s.needCommit() {
 		s.commitPods(committedEndpoints, endpointPodMap)
 	}
 
 	if len(addEndpoints) == 0 && len(removeEndpoints) == 0 {
-		klog.V(4).Infof("No endpoint change for %s/%s, skip syncing NEG. ", s.Namespace, s.Name)
+		s.logger.V(4).Info("No endpoint change. Skip syncing NEG. ", s.Namespace, s.Name)
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/syncers/transaction.go]: Done!

File: pkg/neg/syncers/transaction.go [hunk 4]
```diff
@@ -185,11 +192,10 @@ func (s *transactionSyncer) syncInternalImpl() error {
 	}
 
 	if s.syncer.IsStopped() || s.syncer.IsShuttingDown() {
-		klog.V(4).Infof("Skip syncing NEG %q for %s.", s.NegSyncerKey.NegName, s.NegSyncerKey.String())
+		s.logger.V(4).Info("Skip syncing NEG", "negSyncerKey", s.NegSyncerKey.String())
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/syncers/transaction.go]: Done!

File: cmd/glbc/main.go [hunk 5]
```diff
@@ -339,6 +339,7 @@ func runControllers(ctx *ingctx.ControllerContext) {
 		enableAsm,
 		asmServiceNEGSkipNamespaces,
 		flags.F.EnableEndpointSlices,
+		klog.TODO(), // TODO: Replace this with a top level logger configuration once one is available.
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply cmd/glbc/main.go:342]: Created issue https://github.com/kubernetes/ingress-gce/issues/1761

[the remark this answers is not in the document]
Comment (swetharepakula) [reply cmd/glbc/main.go:342]: I was thinking the other way. Create the issue and add the issue number in the comment.

File: pkg/neg/controller.go [hunk 6]
```diff
@@ -583,7 +592,7 @@ func (c *Controller) mergeVmIpNEGsPortInfo(service *apiv1.Service, name types.Na
 	// Only process ILB services after L4 controller has marked it with v2 finalizer.
 	if !utils.IsSubsettingL4ILBService(service) {
 		msg := fmt.Sprintf("Ignoring ILB Service %s, namespace %s as it does not have the v2 finalizer", service.Name, service.Namespace)
-		klog.Warning(msg)
+		c.logger.Info(msg)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/controller.go:595]: lets keeps this one as info.

File: pkg/neg/controller.go [hunk 7]
```diff
@@ -648,13 +657,13 @@ func (c *Controller) getCSMPortInfoMap(namespace, name string, service *apiv1.Se
...
 		// Find all destination rules that using this service.
-		destinationRules := getDestinationRulesFromStore(c.destinationRuleLister, service)
+		destinationRules := getDestinationRulesFromStore(c.destinationRuleLister, service, c.logger)
 		// Fill all service ports into portinfomap
 		servicePorts := gatherPortMappingFromService(service)
 		for namespacedName, destinationRule := range destinationRules {
 			destinationRulePortInfoMap, err := negtypes.NewPortInfoMapWithDestinationRule(namespace, name, servicePorts, c.namer, false, destinationRule)
 			if err != nil {
-				klog.Warningf("DestinationRule(%s) contains duplicated subset, creating NEGs for the newer ones. %s", namespacedName.Name, err)
+				c.logger.Info("DestinationRule contains duplicated subset, creating NEGs for the newer ones.", "destinationRule", namespacedName.Name, "err", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/controller.go:666]: lets change this to error.

File: pkg/neg/controller.go [hunk 8]
```diff
@@ -769,9 +778,9 @@ func (c *Controller) handleErr(err error, key interface{}) {
 	}
 
 	msg := fmt.Sprintf("error processing service %q: %v", key, err)
-	klog.Errorf(msg)
+	c.logger.Error(nil, fmt.Sprintf(msg))
 	if service, exists, err := c.serviceLister.GetByKey(key.(string)); err != nil {
-		klog.Warningf("Failed to retrieve service %q from store: %v", key.(string), err)
+		c.logger.Info("Failed to retrieve service from store", "service", key.(string), "err", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/controller.go:783]: let's make this error.

File: pkg/neg/controller.go [hunk 9]
```diff
@@ -857,21 +866,21 @@ func (c *Controller) enqueueDestinationRule(obj interface{}) {
...
+func gatherPortMappingUsedByIngress(ings []v1.Ingress, svc *apiv1.Service, logger klog.Logger) negtypes.SvcPortTupleSet {
 	ingressSvcPortTuples := make(negtypes.SvcPortTupleSet)
 	for _, ing := range ings {
 		if utils.IsGLBCIngress(&ing) {
 			utils.TraverseIngressBackends(&ing, func(id utils.ServicePortID) bool {
 				if id.Service.Name == svc.Name && id.Service.Namespace == svc.Namespace {
 					servicePort := translator.ServicePort(*svc, id.Port)
 					if servicePort == nil {
-						klog.Warningf("Port %+v in Service %q not found", id.Port, id.Service.String())
+						logger.Info("Port not found in service", "port", fmt.Sprintf("%+v", id.Port), "service", id.Service.String())
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/controller.go:883]: lets change this to error instead.

File: pkg/neg/manager.go [hunk 10]
```diff
@@ -298,7 +303,7 @@ func (manager *syncerManager) SyncNodes() {
 func (manager *syncerManager) updateZoneMap() bool {
 	zones, err := manager.zoneGetter.ListZones(utils.AllNodesPredicate)
 	if err != nil {
-		klog.Warningf("Unable to list zones: %s", err)
+		manager.logger.Info("Unable to list zones", "err", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/manager.go:306]: lets change this to error.

File: pkg/neg/readiness/poller.go [hunk 11]
```diff
@@ -194,12 +197,12 @@ func (p *poller) processHealthStatus(key negMeta, healthStatuses []*composite.Ne
 
 	for _, healthStatus := range healthStatuses {
 		if healthStatus == nil {
-			klog.Warningf("healthStatus is nil from response %+v", healthStatuses)
+			p.logger.Info("healthStatus is nil from response", "healthStatuses", healthStatuses)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/readiness/poller.go:200]: lets change this to error.

File: pkg/neg/readiness/reflector.go [hunk 12]
```diff
@@ -129,12 +133,12 @@ func (r *readinessReflector) handleErr(err error, key interface{}) {
...
 
 	if r.queue.NumRequeues(key) < maxRetries {
-		klog.V(2).Infof("Error syncing pod %q, retrying. Error: %v", key, err)
+		r.logger.V(2).Info("Error syncing pod. Retrying.", "pod", key, "err", err)
 		r.queue.AddRateLimited(key)
 		return
 	}
 
-	klog.Warningf("Dropping pod %q out of the queue: %v", key, err)
+	r.logger.Info("Dropping pod out of the queue", "pod", key, "err", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/readiness/reflector.go:141]: For this migration, info is good.

File: pkg/neg/syncers/syncer.go [hunk 13]
```diff
@@ -139,7 +142,7 @@ func (s *syncer) Stop() {
 
 func (s *syncer) Sync() bool {
 	if s.IsStopped() {
-		klog.Warningf("NEG syncer for %s is already stopped.", s.NegSyncerKey.String())
+		s.logger.Info("NEG syncer is already stopped.", "negSynckerKey", s.NegSyncerKey.String())
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/syncers/syncer.go:145]: this we can keep as info.

File: pkg/neg/syncers/transaction.go [hunk 14]
```diff
@@ -211,7 +217,7 @@ func (s *transactionSyncer) syncInternalImpl() error {
 			return err
 		}
 		if len(slices) < 1 {
-			klog.Warningf("Endpoint slices for service %s/%s don't exist. Skipping NEG sync", s.Namespace, s.Name)
+			s.logger.Info("Endpoint slices for the service doesn't exist. Skipping NEG sync")
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/syncers/transaction.go:220]: lets change this to error.

File: pkg/neg/syncers/transaction.go [hunk 15]
```diff
@@ -461,7 +467,7 @@ func (s *transactionSyncer) commitPods(endpointMap map[string]negtypes.NetworkEn
 		for _, endpoint := range endpointSet.List() {
 			podName, ok := endpointPodMap[endpoint]
 			if !ok {
-				klog.Warningf("Endpoint %v is not included in the endpointPodMap %v", endpoint, endpointPodMap)
+				s.logger.Info("Endpoint is not included in the endpointPodMap", "endpoint", endpoint, "endpointPodMap", endpointPodMap)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/syncers/transaction.go:470]: let's switch this to error.

File: pkg/neg/syncers/transaction.go [hunk 16]
```diff
@@ -475,23 +481,23 @@ func (s *transactionSyncer) commitPods(endpointMap map[string]negtypes.NetworkEn
 func (s *transactionSyncer) isZoneChange() bool {
 	negCR, err := getNegFromStore(s.svcNegLister, s.Namespace, s.NegSyncerKey.NegName)
 	if err != nil {
-		klog.Warningf("unable to retrieve neg %s/%s from the store: %s", s.Namespace, s.NegName, err)
+		s.logger.Info("unable to retrieve neg from the store", "neg", klog.KRef(s.Namespace, s.NegName), "err", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/syncers/transaction.go:484]: let's switch this to error.

File: pkg/neg/syncers/transaction.go [hunk 17]
```diff
@@ -475,23 +481,23 @@ func (s *transactionSyncer) commitPods(endpointMap map[string]negtypes.NetworkEn
...
 		return false
 	}
 
 	existingZones := sets.NewString()
 	for _, ref := range negCR.Status.NetworkEndpointGroups {
 		id, err := cloud.ParseResourceURL(ref.SelfLink)
 		if err != nil {
-			klog.Warningf("unable to parse selflink %s", ref.SelfLink)
+			s.logger.Info("unable to parse selflink", "selfLink", ref.SelfLink)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/neg/syncers/transaction.go:492]: let's switch this to error.

File: cmd/glbc/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply cmd/glbc/main.go:342]: Done!

[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply cmd/glbc/main.go:342]: Done!

File: pkg/neg/controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/controller.go:595]: Acknowledged.

File: pkg/neg/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/controller.go:666]: Done!

File: pkg/neg/controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/controller.go:883]: Done!

File: pkg/neg/manager.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/manager.go:306]: Done!

File: pkg/neg/readiness/poller.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/readiness/poller.go:200]: Done!

**Comment by spencerhance** [comment]
lgtm

File: pkg/neg/controller.go [hunk 18]
```diff
@@ -792,17 +801,17 @@ func (c *Controller) enqueueEndpointSlice(obj interface{}) {
 	if !ok {
 		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
 		if !ok {
-			klog.Errorf("Unexpected object type: %T, expected cache.DeletedFinalStateUnknown", obj)
+			c.logger.Error(nil, "Unexpected object type, expected cache.DeletedFinalStateUnknown", "objectTypeFound", fmt.Sprintf("%T", obj))
 			return
 		}
 		if endpointSlice, ok = tombstone.Obj.(*discovery.EndpointSlice); !ok {
-			klog.Errorf("Unexpected tombstone object type: %T, expected *discovery.EndpointSlice", obj)
+			c.logger.Error(nil, "Unexpected tombstone object type: %T, expected *discovery.EndpointSlice", "objectTypeFound", fmt.Sprintf("%T", obj))
```
Comment (swetharepakula) [inline pkg/neg/controller.go:808]: remove the format `%T` in the error message.

File: pkg/neg/syncers/transaction.go [hunk 19]
```diff
@@ -541,16 +547,17 @@ func mergeTransactionIntoZoneEndpointMap(endpointMap map[string]negtypes.Network
 
 // logStats logs aggregated stats of the input endpointMap
 func (s *transactionSyncer) logStats(endpointMap map[string]negtypes.NetworkEndpointSet, desc string) {
-	stats := []string{}
+	stats := []interface{}{}
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:550]: nit: var stats []interface{}

**Comment by swetharepakula** [comment]
Thanks for all your work Gaurav. A few more small comments and this should be ready to merge.

File: pkg/neg/controller.go (hunk 18 above)
Comment (gauravkghildiyal) [reply pkg/neg/controller.go:808]: Done!

File: pkg/neg/syncers/transaction.go (hunk 19 above)
Comment (gauravkghildiyal) [reply pkg/neg/syncers/transaction.go:550]: Done!

**Review by swetharepakula — COMMENTED** [review]
Thanks for the change!

/lgtm
/approve

**Pull request merged**